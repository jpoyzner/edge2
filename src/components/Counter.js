import React, { useState } from 'react';
import { connect } from 'react-redux';

function Counter(props) {
  const appCount = useAppCount(props.appCount, props);
  const localCount = useLocalCount(appCount.value, props);

  return (
    <div>
      <div>{props.text + " "}</div>
      <div>
        <span>App Counter = {appCount.value}</span>{' '}
        <button onClick={appCount.onIncrement}>+</button>{' '}
        <button onClick={appCount.onDecrement}>-</button>{' '}
        <button onClick={appCount.incrementIfOdd}>Increment if odd</button>{' '}
        <button onClick={appCount.incrementAsync}>Increment in one second</button>
      </div>
      <div>
        <span>Local Counter = {localCount.value}</span>{' '}
        <button onClick={localCount.onIncrement}>+</button>{' '}
        <button onClick={localCount.onDecrement}>-</button>{' '}
        <button onClick={localCount.incrementIfOdd}>Increment if odd</button>{' '}
        <button onClick={localCount.incrementAsync}>Increment in one second</button>
        <button onClick={localCount.save}>Save to app count</button>
      </div>
    </div>
  );
}

function useAppCount(count, props) {
  return {
    value: count,
    incrementIfOdd() {
      if (count % 2 !== 0) {
        props.increment();
      }
    },
    incrementAsync() {
      setTimeout(() => props.increment(), 1000);
    },
    onIncrement() {
      props.increment()
    },
    onDecrement() {
      props.decrement();
    },
  }
}

function useLocalCount(initialCount, props) {
  const [count, setCount] = useState(initialCount);

  function increment() {
    setCount(count + 1);
  }

  return {
    value: count,
    incrementIfOdd() {
      if (count % 2 !== 0) {
        increment()
      }
    },
    incrementAsync() {
      setTimeout(() => { increment(); }, 1000);
    },
    onIncrement() {
      increment()
    },
    onDecrement() {
      setCount(count - 1);
    },
    save() {
      props.save(count);
    }
  }
}

export default connect(
  (state) => ({ appCount: state.counter }),
  (dispatch) => ({
    increment() {
      dispatch({ type: 'INCREMENT' });
    },
    decrement() {
      dispatch({ type: 'DECREMENT' });
    },
    save(data) {
      dispatch({ type: 'SET', data });
    }
  }),
)(Counter);

