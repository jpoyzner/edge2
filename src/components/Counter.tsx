import React, { SFC, useState } from 'react';
import { connect } from 'react-redux';

interface OwnProps {
  text: string;
}

interface StateProps {
  appCount: number;
}

interface DispatchProps {
  increment(): void;
  decrement(): void;
  save(data: number): void;
}

type Props = StateProps & DispatchProps & OwnProps;

const Counter: SFC<Props> = (props) => {
  const appCount: AppCountState = useAppCount(props.appCount, props);
  const localCount: LocalCountState = useLocalCount(appCount.value, props);

  return (
    <div>
      <div>{props.text + " "}</div>
      <div>
        <span>App Counter = {appCount.value}</span>{' '}
        <button onClick={appCount.onIncrement}>+</button>{' '}
        <button onClick={appCount.onDecrement}>-</button>{' '}
        <button onClick={appCount.incrementIfOdd}>Increment if odd</button>{' '}
        <button onClick={appCount.incrementAsync}>Increment in one second</button>
      </div>
      <div>
        <span>Local Counter = {localCount.value}</span>{' '}
        <button onClick={localCount.onIncrement}>+</button>{' '}
        <button onClick={localCount.onDecrement}>-</button>{' '}
        <button onClick={localCount.incrementIfOdd}>Increment if odd</button>{' '}
        <button onClick={localCount.incrementAsync}>Increment in one second</button>
        <button onClick={localCount.save}>Save to app count</button>
      </div>
    </div>
  );
}

interface AppCountState {
  value: number;
  incrementIfOdd(): void;
  incrementAsync(): void;
  onIncrement(): void;
  onDecrement(): void;
}

function useAppCount(count: number, props: DispatchProps): AppCountState {
  return {
    value: count,
    incrementIfOdd() {
      if (count % 2 !== 0) {
        props.increment();
      }
    },
    incrementAsync() {
      setTimeout(() => props.increment(), 1000);
    },
    onIncrement() {
      props.increment()
    },
    onDecrement() {
      props.decrement();
    },
  }
}

interface LocalCountState {
  value: number;
  incrementIfOdd(): void;
  incrementAsync(): void;
  onIncrement(): void;
  onDecrement(): void;
  save(): void;
}

function useLocalCount(initialCount: number, props: DispatchProps): LocalCountState {
  const [count, setCount] = useState(initialCount);

  function increment() {
    setCount(count + 1);
  }

  return {
    value: count,
    incrementIfOdd() {
      if (count % 2 !== 0) {
        increment()
      }
    },
    incrementAsync() {
      setTimeout(() => { increment(); }, 1000);
    },
    onIncrement() {
      increment()
    },
    onDecrement() {
      setCount(count - 1);
    },
    save() {
      props.save(count);
    }
  }
}

export default connect(
  (state: Map<string, any>, ownProps: OwnProps): StateProps => ({
    appCount: state.get('counter'),
  }),
  (dispatch, ownProps: OwnProps): DispatchProps => ({
    increment() {
      dispatch({ type: 'INCREMENT' });
    },
    decrement() {
      dispatch({ type: 'DECREMENT' });
    },
    save(data: number) {
      dispatch({ type: 'SET', data });
    }
  }),
)(Counter);
