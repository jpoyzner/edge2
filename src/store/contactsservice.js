import Store from '../store/store';

const JSON_SERVER_PORT = 4000;
const CONTACTS_URL = `http://localhost:${JSON_SERVER_PORT}/contacts`;

export default store => next => action => {
  next(action);
	switch (action.type) {
  	case 'getContacts':
  		getContacts().then((response) => {
        next({ type: 'setContacts', data: response });
      }) 

    	break;
    case 'addContact':
      getContacts().then((response) => {
        fetch(CONTACTS_URL, {
          method: 'post',
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            id: response.length,
            name: 'Jeff Poyzner',
            number: `+1${action.data.number}`,
            context: 'Edge2',
          })
        })
        .then(() => {
          Store.dispatch({type: 'getContacts'});
        });
      }) ;

      break;
    default: return false;
	}
};

function getContacts() {
  return new Promise((resolve) => {
    fetch(CONTACTS_URL)
      .then(res => res.json())
      .then(response => {
        //console.log('Success:', JSON.stringify(response));
        resolve(response);
      })
      .catch(error => {
        console.error('Error:', error);
      });
  });
}