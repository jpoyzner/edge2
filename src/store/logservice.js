export default store => next => async action => {
  next(action);
	switch (action.type) {
  	case 'GETLOGS':
      try {
        const res = await fetch('/logs/' + action.query);
        const response = await res.json();
        next({type: 'GOTLOGS', data: response});
      } catch(err) {
        console.error(err);
      }

    	break;
    default: return false;
	}
};