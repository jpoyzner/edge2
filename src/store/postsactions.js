export default (state, action) => {
  switch (action.type) {
	 	case 'setPosts':
	 		state.posts = action.data;
	 		return state;
	  default: return false;
	}
};
