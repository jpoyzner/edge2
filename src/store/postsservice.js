export default store => next => action => {
  next(action);
	switch (action.type) {
  	case 'getPosts':
  		fetch('https://jsonplaceholder.typicode.com/posts')
        .then(res => res.json())
        .then(response => {
          //console.log('Success:', JSON.stringify(response));
          next({type: 'setPosts', data: response});
        })
        .catch(error => console.error('Error:', error));	      

    	break;
    default: return false;
	}
};