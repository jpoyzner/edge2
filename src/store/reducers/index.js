import { combineReducers } from 'redux-immutable';
import counter from './counter';
import logs from './logs';
import posts from './posts';
import todos from './todos';
import contacts from './contacts';

export default combineReducers({
  counter,
  logs,
  posts,
  todos,
  contacts,
})