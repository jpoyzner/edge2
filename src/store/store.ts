import { configureStore } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import counter from './reducers/Counter';
import logs from './reducers/Logs';
import posts from './reducers/Posts';
import todos from './reducers/Todos';
import contacts from './reducers/Contacts';

const store = configureStore({
  reducer: {
    counter,
    logs,
    posts,
    todos,
    contacts,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  devTools: process.env.NODE_ENV !== 'production',
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, AnyAction>;
export default store;