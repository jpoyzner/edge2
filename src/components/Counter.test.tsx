import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import store from '../store/store';
import Counter from './Counter';
import { ReactWrapper } from 'enzyme';

test('Counter counts', () => {
  const text: string = "blahblah";

  const app: ReactWrapper = mount(
    <Provider store={store}>
      <Counter text={text} />
    </Provider>
  );

  const component: ReactWrapper = app.find('Counter');
  expect(component).toMatchSnapshot();

  expect(component.find('.jp-counter-text').text()).toEqual(text);

  const props: any = component.props();
  props.increment();
  props.increment();
  props.increment();
  expect(component.find('.jp-counter-count').text()).toEqual('App Counter = 3');
});

