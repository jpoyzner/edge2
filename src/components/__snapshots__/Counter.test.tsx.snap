// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter counts 1`] = `ReactWrapper {}`;
